File: LeetCode/Python3/Hard/4. Median of Two Sorted Arrays/0004-median-of-two-sorted-arrays.py
import sys
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        m = len(nums1)
        n = len(nums2)
        start = 0
        end = m
        while start <= end:
            partition_nums1 = (start + end) // 2
            partition_nums2 = (m + n + 1) // 2 - partition_nums1
            maxLeftNums1 = -sys.maxsize if partition_nums1 == 0 else nums1[partition_nums1 - 1]
            minRightNums1 = sys.maxsize if partition_nums1 == m else nums1[partition_nums1]
            maxLeftNums2 = -sys.maxsize if partition_nums2 == 0 else nums2[partition_nums2 - 1]
            minRightNums2 = sys.maxsize if partition_nums2 == n else nums2[partition_nums2]
            if maxLeftNums1 <= minRightNums2 and maxLeftNums2 <= minRightNums1:
                if (m + n) % 2 == 0:
                    return (max(maxLeftNums1, maxLeftNums2) + min(minRightNums1, minRightNums2)) / 2
                else:
                    return max(maxLeftNums1, maxLeftNums2)
            elif maxLeftNums1 > minRightNums2:
                end = partition_nums1 - 1
            else:
                start = partition_nums1 + 1

File: LeetCode/C++/Hard/4. Median of Two Sorted Arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() < nums2.size()) {
            return findMedianSortedArrays(nums2, nums1);
        }
        int low = 0;
        int high = nums2.size() * 2;
        while (low <= high) {
            int midTwo = (low + high) / 2;
            int midOne = nums1.size() + nums2.size() - midTwo;
            double leftOne = (midOne == 0) ? INT_MIN : nums1[(midOne - 1) / 2];	
            double leftTwo = (midTwo == 0) ? INT_MIN : nums2[(midTwo - 1) / 2];
            double rightOne = (midOne == nums1.size() * 2) ? INT_MAX : nums1[(midOne) / 2];
            double rightTwo = (midTwo == nums2.size() * 2) ? INT_MAX : nums2[(midTwo) / 2];
            if (leftOne > rightTwo) {
                low = midTwo + 1;
            } else if (leftTwo > rightOne) {
                high = midTwo - 1;
            } else {
                return (max(leftOne, leftTwo) + min(rightOne, rightTwo)) / 2;	
            }
        }
        return -1;
    }
};